File: src/main/java/com/hazza/LeetCodeByJava/SegmentOne/No9PalindromeNumber.java
package com.hazza.LeetCodeByJava.SegmentOne;

/**
 * https://leetcode.com/problems/palindrome-number/#/description
 */
public class No9PalindromeNumber {
    public boolean isPalindrome(int x) {
        if (x < 0 || (x != 0 && x % 10 == 0)) return false;   //舍弃负数和个位是0的数字
        int y = 0;
        while (x > y) { //防止溢出
            y = y * 10 + x % 10;
            x /= 10;
        }
        return (x == y || x == y / 10); //前一个条件为了x是偶数，后一个条件为了x使奇数
    }
}

File: src/main/python/com/hazza/leetcode_by_python/SegmentOne/No9_Palindrome_Number.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/7 21:43
# https://leetcode.com/problems/palindrome-number/


class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0 or x != 0 and x % 10 == 0:
            return False
        y = 0
        while x > y:
            x, tail = divmod(x, 10)
            y = y * 10 + tail

        return x == y or int(y / 10) == x


if __name__ == '__main__':
    solution = Solution()
    assert solution.isPalindrome(121) == True
    assert solution.isPalindrome(-121) == False
    assert solution.isPalindrome(10) == False
